src/commands/talk.ts

```diff
@@ -0,0 +1,224 @@
+	async execute(context) {
+		const options = context.interaction.options as CommandInteractionOptionResolver;
+
+		// this shouldn't happen, but it's good to have error checking anyway
+		if (options.data.length === 0) {
+			throw new Error('No options provided');
+		}
+
+		const message = options.getString('message', true);
+		const speakerNum = options.getInteger('speaker') ?? undefined;
+
+		await speak(message, speakerNum, context);
+	},
+};
+
+/**
+ * This method abstracts-out the functionality of the "talk" command, allowing it to be used
+ * by either slash commands or context commands
+ * @param message The message to speak
+ * @param speakerNum Which speaker to use (can be undefined)
+ * @param context The context of the interaction (either TextInputCommandContext or MessageContextCommandContext)
+ */
+export async function speak(
+	message: string,
+	speakerNum: Speaker | undefined,
+	{ channel, client, prepareForLongRunningTasks, reply }: BaseCommandContext
+): Promise<void> {
+	// this also shouldn't happen, but better safe than sorry
+	if (!channel) {
+		throw new Error('No channel');
+	}
+
+	// this could happen if the user uses the context menu command on an empty message
+	if (!message || message === '') {
+		throw new Error('Message cannot be empty');
+	}
+
+	// generating the audio can take a while, so make sure Discord doesn't think we died
+	await prepareForLongRunningTasks(false);
+
+	// generate the audio and store it in a generic buffer
+	log('generating audio');
+	const wavData: Buffer = await say(message, {
+		Speaker: speakerNum,
+	});
+
+	// if the command was given in a voice chat, speak it
+	if (channel.type === ChannelType.GuildVoice) {
+		// make sure we have the right permissions to talk
+		const permissions = channel.permissionsFor(client.user);
+		if (!permissions) {
+			throw new Error('Could not fetch permissions for channel');
+		}
+		if (!permissions.has(PermissionFlagsBits.Connect)) {
+			throw new Error('Missing required permissions to connect to voice channel');
+		}
+		if (!permissions.has(PermissionFlagsBits.Speak)) {
+			throw new Error('Missing required permissions to speak in voice channel');
+		}
+
+		// check to see if we're already connected to this channel
+		// a channel can only have one connection & one player at a time!
+		if (getVoiceConnection(channel.guild.id)) {
+			throw new Error('Already talking in channel\n\nPlease try again later');
+		}
+
+		// store the audio buffer in a temporary .wav file to pass to discordjs/voice
+		const tempFile = fileSync({ prefix: 'dectalk', postfix: '.wav' });
+		writeFileSync(tempFile.name, wavData);
+		const stream = createReadStream(tempFile.name);
+
+		// create an audio resource from the temporary file
+		let resource: AudioResource;
+		try {
+			resource = createAudioResource(stream);
+		} catch (error) {
+			const error_ = error as Error;
+			// For this error, provide a more useful message
+			if (error_.message === 'FFmpeg/avconv not found!') {
```

Comment:
Avoid typecasts, unless there's no other way and assumptions _must_ be made.

I hope the thrown error here is an `Error` or `Error` subclass. If not, then this `instanceof` check should be replaced with something more complicated to assert that `message` exists.

```suggestion
			// For this error, provide a more useful message
			if (error instanceof Error && error.message === 'FFmpeg/avconv not found!') {
```